internal/plugins.go

```diff
@@ -0,0 +1,281 @@
+		Name          string                 `json:"name"`       // descriptive plugin name
+		Encodings     []string               `json:"encodings"`  // list of supported encodings
+		EventCount    int                    `json:"eventCount"` // number of events that have passed through this plugin
+		routes        []*Route               // list of routes using this plugin instance
+		inputChannel  chan *Event            // event channel on which plugin receives the next event
+		outputChannel chan *Event            // event channel to which plugin forwards current event to
+		done          chan bool              // done channel
+		filterer      Filterer               // an instance of the appropriate filterer
+	}
+	FilterPlugin struct {
+		Plugin
+	}
+	DebugInputPlugin struct {
+		Plugin
+		IntervalMs int
+		Rounds     int
+		Payload    interface{}
+	}
+	DebugOutputPlugin struct {
+		Plugin
+	}
+)
+
+//
+// generic plugin features
+//
+
+func (plgn *Plugin) Hash(ctx context.Context) string {
+	str := ""
+	// distinguish different plugin types
+	str += plgn.Type
+	// distinguish different configurations
+	if plgn.Params != nil {
+		buf, _ := json.Marshal(plgn.Params)
+		str += string(buf)
+	}
+	// distinguish input and output plugins
+	str += plgn.Mode
+	// optionally distinguish by org and app here as well
+	hash := fmt.Sprintf("%x", md5.Sum([]byte(str)))
+	return hash
+}
+
+func (plgn *Plugin) Validate(ctx context.Context) error {
+	return nil
+}
+
+func (plgn *Plugin) Initialize(ctx context.Context) error {
+	return nil
+}
+
+func (plgn *Plugin) String() string {
+	buf, _ := json.Marshal(plgn)
+	return string(buf)
+}
+
+func (plgn *Plugin) GetConfig() *Plugin {
+	return plgn
+}
+
+func (plgn *Plugin) GetInputChannel() chan *Event {
+	return plgn.inputChannel
+}
+
+func (plgn *Plugin) GetOutputChannel() chan *Event {
+	return plgn.outputChannel
+}
+
+func (plgn *Plugin) GetRouteCount() int {
+	if plgn.routes != nil {
+		return len(plgn.routes)
+	}
+	return 0
+}
+
+func (plgn *Plugin) GetEventCount() int {
+	return plgn.EventCount
+}
+
+func (plgn *Plugin) DoSync(ctx context.Context, event *Event) error {
+	return new(UnworthyPluginError)
+}
+
+func (plgn *Plugin) DoAsync(ctx context.Context) {
+}
+
+func (plgn *Plugin) Close(ctx context.Context) {
+	log.Error().Msg("sending done signal for " + plgn.Mode + " " + plgn.Type + " " + plgn.Hash(ctx))
+	plgn.done <- true
+}
+
+//
+// debug input plugin
+//
+
+func (dip *DebugInputPlugin) DoAsync(ctx context.Context) {
+	done := false
+	go func() {
+		<-dip.done
+		done = true
+	}()
+	go func() {
+		if dip.Payload == nil {
+			log.Error().Msg("no payload configured for debug input plugin " + dip.Hash(ctx))
+			return
+		}
+		for {
+			if dip.Rounds > 0 && dip.EventCount >= dip.Rounds {
+				return
+			}
+			time.Sleep(time.Duration(dip.IntervalMs) * time.Millisecond)
+			if done {
+				break
+			}
+			event := NewEvent(ctx, dip, dip.Payload)
```

Comment:
The ctx here is the global context. We need to figure out where to generate sub context. Is it inside the NewEvent or do we have the sub context even before we call the NewEvent constructor

Reply:
i think the input plugin is responsible for new context generation, but yes it could delegate that to new plugin